server/schedule/region_splitter.go

```diff
@@ -79,62 +87,59 @@ func (r *RegionSplitter) SplitRegions(splitKeys [][]byte, retryLimit int) (int,
 	return 100 - len(unprocessedKeys)*100/len(splitKeys), returned
 }
 
-func (r *RegionSplitter) splitRegionsByKeys(splitKeys [][]byte, newRegions map[uint64]struct{}) [][]byte {
-	//TODO: support batch limit
-	groupKeys, unProcessedKeys := r.groupKeysByRegion(splitKeys)
-	for regionID, keys := range groupKeys {
-		region := r.cluster.GetRegion(regionID)
-		if region == nil {
-			unProcessedKeys = append(unProcessedKeys, keys...)
-			continue
-		}
-		if !r.checkRegionValid(region) {
-			unProcessedKeys = append(unProcessedKeys, keys...)
-			continue
-		}
-		err := r.handler.SplitRegionByKeys(region, keys)
+func (r *RegionSplitter) splitRegionsByKeys(ctx context.Context, splitKeys [][]byte, newRegions map[uint64]struct{}) [][]byte {
+	validGroups := r.groupKeysByRegion(splitKeys)
+	for key, group := range validGroups {
+		err := r.handler.SplitRegionByKeys(group.region, group.keys)
 		if err != nil {
-			unProcessedKeys = append(unProcessedKeys, keys...)
+			delete(validGroups, key)
 			continue
 		}
-		// TODO: use goroutine to run watchRegionsByKeyRange asynchronously
-		// TODO: support configure timeout and interval
-		splittedRegionsID := r.handler.WatchRegionsByKeyRange(region.GetStartKey(), region.GetEndKey(),
-			keys, time.Minute, 100*time.Millisecond)
-		for key := range splittedRegionsID {
-			newRegions[key] = struct{}{}
-		}
 	}
-	return unProcessedKeys
+	wg := &sync.WaitGroup{}
+	results := newSplitKeyResults()
+	for _, groupKeys := range validGroups {
+		wg.Add(1)
+		// TODO use recovery to handle error/panic
+		go r.handler.WatchRegionsByKeyRange(ctx, groupKeys.region.GetStartKey(), groupKeys.region.GetEndKey(),
```

Comment:
How about only passing groupKeys to reduce the number of parameters?

Reply:
updated.